core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala

```diff
@@ -0,0 +1,369 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.cluster
+
+import java.util.Properties
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicInteger}
+import java.util.concurrent.locks.ReadWriteLock
+import java.util.concurrent.{ArrayBlockingQueue, Executors, Future, TimeUnit}
+
+import kafka.api.{ApiVersion, LeaderAndIsr}
+import kafka.log._
+import kafka.server._
+import kafka.server.checkpoints.OffsetCheckpoints
+import kafka.utils._
+import org.apache.kafka.common.message.LeaderAndIsrRequestData.LeaderAndIsrPartitionState
+import org.apache.kafka.common.{MetricName, TopicPartition}
+import org.apache.kafka.common.metrics.stats.{Avg, Max}
+import org.apache.kafka.common.metrics.Metrics
+import org.apache.kafka.common.record.{MemoryRecords, SimpleRecord}
+import org.apache.kafka.common.utils.Utils
+import org.junit.Assert.{assertEquals, assertTrue}
+import org.junit.{After, Before, Test}
+import org.mockito.ArgumentMatchers
+import org.mockito.Mockito.{mock, when}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+
+/**
+ * Verifies that slow appends to log don't block request threads processing replica fetch requests.
+ *
+ * Test simulates:
+ *   1) Produce request handling by performing append to log as leader.
+ *   2) Replica fetch request handling by processing update of ISR and HW based on log read result.
+ *   3) Some tests also simulate a scheduler thread that checks or updates ISRs when replica falls out of ISR
+ */
+class PartitionLockTest extends Logging {
+
+  val numReplicaFetchers = 2
+  val numProducers = 3
+  val numRecordsPerProducer = 5
+  // Use artificially large delays to enable automated testing of locking delays
+  val appendDelayMs = 300
+  val isrCheckDelayMs = 10
+  val isrUpdateDelayMs = 50
+
+  val mockTime = new MockTime()
```

Comment:
The mix of mock time and system time in these tests is a little hard to follow. Maybe a brief comment of explanation would help.

Reply:
The tests were originally written to get some timings. I converted that to an integration test, but actually, that is probably not that useful to detect regressions since the times used in the test need to be large enough to avoid flaky tests. I have rewritten the tests to use semaphores to verify the conditions without relying on actual time. Hopefully the code is easier to understand and also more reliable.